op_crates/crypto/01_crypto.js

```diff
@@ -1,25 +1,206 @@
+
+    constructor(definitions) {
+      this.#nameIndex = Object.create(null);
+      this.#definitions = Object.create(null);
+      for (const [name, definition] of Object.entries(definitions)) {
+        this.#nameIndex[name.toUpperCase()] = name;
+        this.#definitions[name] = definition;
+      }
+    }
+
+    getDefinition(name) {
+      return this.#definitions[name];
+    }
+
+    normalizeName(providedName) {
+      const upperCaseName = toUpperCase.call(providedName);
+      return this.#nameIndex[upperCaseName];
+    }
+  }
+
+  const supportedAlgorithms = {
+    digest: new RegisteredAlgorithmsContainer({
+      "SHA-1": {},
+      "SHA-256": {},
+      "SHA-384": {},
+      "SHA-512": {},
+    }),
+  };
+
+  const digestByteLengths = {
+    "SHA-1": 20,
+    "SHA-256": 32,
+    "SHA-384": 48,
+    "SHA-512": 64,
+  };
+
+  function normalizeAlgorithm(algorithm, registeredAlgorithms) {
+    let alg;
+    if (typeof algorithm === "string") {
+      alg = { name: algorithm };
+    }
+    if (typeof algorithm === "object" && algorithm !== null) {
+      if (typeof algorithm.name !== "string") {
+        throw new TypeError("Algorithm name is missing or not a string");
+      }
+      const algorithmName = registeredAlgorithms.normalizeName(algorithm.name);
+      if (algorithmName === undefined) {
+        throw new DOMException(
+          "Unrecognized algorithm name",
+          "NotSupportedError",
+        );
+      }
+      const definition = registeredAlgorithms.getDefinition(algorithmName);
+      // TODO: Validate the algorithm data based on the definition
+      alg = { name: algorithmName };
+    } else {
+      throw new TypeError("Argument 1 must be an object or a string");
+    }
+    return alg;
+  }
+
+  const subtle = {
+    async decrypt(algorithm, key, data) {
+      await Promise.resolve();
+      throw new Error("Not implemented");
+    },
+    async deriveBits(algorithm, baseKey, length) {
+      await Promise.resolve();
+      throw new Error("Not implemented");
+    },
+    async deriveKey(
+      algorithm,
+      baseKey,
+      derivedKeyType,
+      extractable,
+      keyUsages,
+    ) {
+      await Promise.resolve();
+      throw new Error("Not implemented");
+    },
+    async digest(algorithm, data) {
+      const alg = normalizeAlgorithm(algorithm, supportedAlgorithms.digest);
+      // TODO: validate that data is indeed a BufferSource
+      // Note: we need to copy the data because the specs require it.
+      // It's unclear whether we are better doing it here or on the Rust side.
+      const input = data.slice(0);
+      const output = new ArrayBuffer(digestByteLengths[alg.name]);
+      await core.jsonOpAsync(
+        "op_crypto_subtle_digest",
+        alg,
+        // TODO: Check that chugging the input and output directly to the
+        // constructor of Uint8Array doesn't cause issues of length, nor
+        // does it perform any useless data copy.
+        new Uint8Array(input),
```

Comment:
`new TypedArray(TypedArray)` copies unlike `new TypedArray(ArrayBuffer)`
length issues will only happen if you pass non u8 TypedArray, but it can be easily fixed by operating on underlaying ArrayBuffer
```suggestion
        new Uint8Array(input.buffer, input.byteOffset, input.byteLength),
```

given that spec allows passing ArrayBuffer you probably will want to do something like this

```js
function validateBufferSource(data) {
  if (data instanceof ArrayBuffer) return new Uint8Array(data);
  if (ArrayBuffer.isView(data)) return new Uint8Array(data.buffer, data.byteOffset, data.byteLength);

  throw new TypeError("The provided value is not of type '(ArrayBuffer or ArrayBufferView)'");
}

const u8array = validateBufferSource(data); // .slice() for copy
```

Reply:
Thank you very much for the info! It was very helpful.